
# COJAC - CoOccurrence adJusted Analysis and Calling


## Description

The _cojac_ package comprises a set of command-line tools to analyse co-occurrence of mutations on amplicons. It is useful, for example, for finding viral variants of concern in environmental samples, and has been designed to scan for the SARS-CoV-2 variants [B.1.1.7](https://virological.org/t/preliminary-genomic-characterisation-of-an-emergent-sars-cov-2-lineage-in-the-uk-defined-by-a-novel-set-of-spike-mutations/563) and [501.V2](https://doi.org/10.1101%2F2020.12.21.20248640) in wastewater samples.

The analysis requires the whole amplicon to be covered by sequencing read pairs. It currently works at the level of aligned reads, but [we plan](#upcoming-features) to be able to adjust confidence scores based on local (window) haplotypes (as generated, e.g., by [ShoRAH](https://github.com/cbg-ethz/shorah), [doi:10.1186/1471-2105-12-119](https://doi.org/10.1186/1471-2105-12-119)).

## Usage

Here are the available command-line tools:

| command                              | purpose |
| :----------------------------------- | :------ |
| [`cooc-mutbamscan`](cooc-mutbamscan) | scan an alignment BAM/CRAM/SAM file for mutation co-occurrences and output a JSON or YAML file |
| [`cooc-colourmut`](cooc-colourmut)   | display a JSON or YAML file as a coloured output on the terminal |
| [`cooc-pubmut`](cooc-pubmut)         | render a JSON or YAML file to a table as in the publication |

Use option `-h` / `--help` to see available command-line options:

```
cooc-mutbamscan --help
usage: cooc-mutbamscan [-h] (-s TSV | -a BAM/CRAM [BAM/CRAM ...]) [-p PATH] [-j JSON] [-y YAML] [-t TSV] [-d]

scan amplicon (covered by long read pairs) for mutation cooccurrence

optional arguments:
-h, --help show this help message and exit
-s TSV, --samples TSV
V-pipe samples list tsv
-a BAM/CRAM [BAM/CRAM ...], --alignments BAM/CRAM [BAM/CRAM ...]
alignment files
-p PATH, --prefix PATH
V-pipe work directory prefix for where to look at align files when using TSV samples list
-j JSON, --json JSON output results to as JSON file
-y YAML, --yaml YAML output results to as yaml file
-t TSV, --tsv TSV output results to as (raw) tsv file
-d, --dump dump the python object to the terminal
```

```
cooc-colourmut --help
usage: cooc-colourmut [-h] (-j JSON | -y YAML)

print coloured pretty table on terminal

optional arguments:
-h, --help show this help message and exit
-j JSON, --json JSON results generated by mutbamscan
-y YAML, --yaml YAML results generated by mutbamscan

see cooc-pubmut for a CSV file that can be imported into an article
```

```
cooc-pubmut --help
usage: cooc-pubmut [-h] (-j JSON | -y YAML) [-o CSV] [-e | -x] [-q]

make a pretty table

optional arguments:
-h, --help show this help message and exit
-j JSON, --json JSON results generated by mutbamscan
-y YAML, --yaml YAML results generated by mutbamscan
-o CSV, --output CSV name of (pretty) csv file to save the table into
-e, --escape use escape characters for newlines
-x, --excel use a semi-colon ';' instead of a comma ',' in the comma-separated-files as required by Microsoft Excel
-q, --quiet Run quietly: do not print the table

you need to open the CSV in a spreadsheet that understands linebreaks
```

## Howto

### Input data requirements

Analysis needs to be performed on SARS-CoV-2 samples sequenced using [ARTIC V3 protocol](https://doi.org/10.17504/protocols.io.bibtkann) (which produces ~400bp long amplicons), and sequenced with read settings that covers the totality of an amplicon (e.g.: paired end sequencing with read length 250).
NOTE: this analysis method cannot work on read length much shorter than the amplicons (e.g.: it will not give reliable results for a read-length of 50).

### Collect the co-occurrence data

There are currently two modes to collect the data about co-occurring mutations in reads: analysing stand-alone BAM/CRAM/SAM alignment files, or analysing the output of a cohort analysed with [V-pipe](https://cbg-ethz.github.io/V-pipe/) ([doi:10.1093/bioinformatics/btab015](https://doi.org/10.1093/bioinformatics/btab015)).

#### Standalone files

Provide a list of BAM files using the `-a` / `--alignment` option. Run:

```bash
cooc-mutbamscan -a sam1.bam sam2.bam -j cooc-test.json
```

> **Note:** you can also use the `-y` / `--yaml` option to write to a YAML file instead of a JSON.

#### Analyzing a cohort with V-pipe

You can learn how to analyse _fastq.gz_ files with V-pipe with this tutorial:

 - https://cbg-ethz.github.io/V-pipe/tutorial/sars-cov2/ ([video tutorial](https://youtu.be/pIby1UooK94))

Run:

```bash
cooc-mutbamscan -t work/samples.tsv -p work/samples/ -j cooc-test.json
```

### Display data on terminal

The default `-d` / `--dump` option of `cooc-mutbamscan` is not a very user-friendly experience to display the data. You can instead pass a JSON or YAML file to the display script. Run:

```bash
cooc-colourmut -j cooc-test.json
```

<img src='data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" viewBox="0, 0, 1436.63, 117.45" font-family="SauceCodePro Nerd Font, Source Code Pro, Courier" font-size="14"><g fill="%23e0e0e0"><rect x="0" y="0" width="1436.63" height="117.45" fill="%232c3e50"/><text x="84.01" y="14.55">72_UK                      78_UK                      92_UK                      93_UK                      76_SA                      77_EU</text><text x="84.01" y="33.55">21765-21770&%23x394;,21991-21993&%23x394;  23604A,23709T              27972T,28048T,28111G       28111G,28280-28280-&%23x3E;CTA    23012A,23063T              23403G</text><text x="0" y="52.55">sample:       cov:     mut:   frq/%:     cov:     mut:   frq/%:     cov:     mut:   frq/%:     cov:     mut:   frq/%:     cov:     mut:   frq/%:     cov:     mut:   frq/%:</text><text x="0" y="71.55">sam1.bam       809</text><text x="201.63" y="71.55" font-weight="bold" fill="%232ecc71">158  19.53%2</text><text x="352.86" y="71.55">452</text><text x="445.27" y="71.55" font-weight="bold" fill="%232ecc71">2   0.44%2</text><text x="579.69" y="71.55">400</text><text x="663.71" y="71.55" font-weight="bold" fill="%232ecc71">89  22.25%3</text><text x="806.53" y="71.55">758</text><text x="882.14" y="71.55" font-weight="bold" fill="%232ecc71">344  45.38%2</text><text x="1024.97" y="71.55">1090</text><text x="1108.98" y="71.55" font-weight="bold" fill="%23e9bd0e">377    (1/2)</text><text x="1260.21" y="71.55">371</text><text x="1335.82" y="71.55" font-weight="bold" fill="%232ecc71">371 100.00%1</text><text x="0" y="90.55">sam2.bam      1121</text><text x="201.63" y="90.55" font-weight="bold" fill="%23ce4435">N/A      N/A</text><text x="352.86" y="90.55">255</text><text x="436.87" y="90.55" font-weight="bold" fill="%23e9bd0e">52    (1/2)</text><text x="579.69" y="90.55">432</text><text x="663.71" y="90.55" font-weight="bold" fill="%232ecc71">58  13.43%3</text><text x="806.53" y="90.55">958</text><text x="882.14" y="90.55" font-weight="bold" fill="%232ecc71">142  14.82%2</text><text x="1024.97" y="90.55">1005</text><text x="1108.98" y="90.55" font-weight="bold" fill="%23ce4435">N/A      N/A</text><text x="1251.8" y="90.55">1615</text><text x="1327.42" y="90.55" font-weight="bold" fill="%232ecc71">1615 100.00%1</text></g></svg>' alt="[terminal screen shot]" />

### Render table for publication

And now, let’s go beyond our terminal and produce a table that can be included in a publication (see bibliography below for concrete example). Run:

```bash
cooc-pubmut -j cooc-test.json -o cooc-output.tsv
```

> **Note:**
> - you can also output to comma-separated table (`-o cooc-output.csv`)
> - Microsoft Excel requires using option `-x`/`--excel` (using semi-colon instead of comma in comma-separated-value files). Some versions can also open TSV (but not the Office 365 web app).

You need to open the table with a spread-sheet that can understand line breaks, such as [LibreOffice Calc](https://www.libreoffice.org/discover/calc/), [Google Docs Spreadsheet](https://www.google.com/sheets/about/) or, using special options (see above), [Microsoft Excel](https://www.microsoft.com/en-us/microsoft-365/excel).

|          | 72_UK               | 78_UK            | 92_UK              | 93_UK               | 76_SA             | 77_EU                  |
| :------- | ------------------: | ---------------: | -----------------: | ------------------: | ----------------: | ---------------------: |
| sam1.bam | 158 / 809<br>19.53% | 2 / 452<br>0.44% | 89 / 400<br>22.25% | 344 / 758<br>45.38% | 0 / 1090<br>0.00% | 371 / 371<br>100.00%   |
| sam2.bam | 0 / 1121<br> 0.00%  | 0 / 255<br>0.00% | 58 / 432<br>13.43% | 142 / 958<br>14.82% | 0 / 1005<br>0.00% | 1615 / 1615<br>100.00% |


It is also possible to use the software [_pandoc_](https://pandoc.org/) to further convert the CSV to other formats. Run:

```bash
cooc-pubmut -j cooc-test.json -o cooc-output.csv
pandoc cooc-output.csv -o cooc-output.pdf
pandoc cooc-output.csv -o cooc-output.html
pandoc cooc-output.csv -o cooc-output.md
```

## Installation

We recommend using bioconda software repositories for easy installation.
You can find instruction to setup your bioconda environment at the following address:

 - https://bioconda.github.io/user/install.html

If you use [V-pipe’s `quick_install.sh`](https://cbg-ethz.github.io/V-pipe/tutorial/sars-cov2/#install-v-pipe), it will set up an environment that you can activate, e.g.:

```bash
bash quick_install.sh -b sars-cov2 -p testing -w work
. ./testing/miniconda3/bin/activate
```

### Prebuilt package

A prebuilt package will be available shortly in the bioconda repository. Come back in the following days for an update.

### Dependencies

If you want to install the software yourself, you can see the list of dependencies in [`conda_cojac_env.yaml`](conda_cojac_env.yaml).

We recommend using conda to install them:

```bash
conda env create -f conda_cojac_env.yaml
conda activate cojac
```

## Upcoming features

- [ ] bioconda package
- [ ] further jupyter and rstudio code from the publication
- [ ] Move hard-coded amplicons to BED input file
- [ ] Move hard-coded mutations to YAML configuration

Long term goal:

- [ ] Integration with ShoRAH amplicons

## Contributions

#### Package developers:

- [David Dreifuss ![orcid ]](https://orcid.org/0000-0002-5827-5387), [![github]](https://github.com/dr-david)
- [Ivan Topolsky ![orcid]](https://orcid.org/0000-0002-7561-0810), [![github]](https://github.com/dryak)

#### Additional notebooks:

 - [Katharina Jahn ![orcid]](https://orcid.org/0000-0002-6983-4615), [![github]](https://github.com/jahnka)

#### Corresponding author:

 - [Niko Beerenwinkel ![orcid]](https://orcid.org/0000-0002-0573-6119)

[github]: https://cbg-ethz.github.io/V-pipe/img/mark-github.svg
[orcid]: https://orcid.org/sites/default/files/images/orcid_16x16.png

## Citation

If you use this software in your research, please cite:

- Katharina Jahn, David Dreifuss, Ivan Topolsky, Anina Kull, Pravin Ganesanandamoorthy, Xavier Fernandez-Cassi, Carola Bänziger, Elyse Stachler, Lara Fuhrmann, Kim Philipp Jablonski, Chaoran Chen, Catharine Aquino, Tanja Stadler, Christoph Ort, Tamar Kohn, Timothy R. Julian, Niko Beerenwinkel

  "*Detection of SARS-CoV-2 variants in Switzerland by genomic analysis of wastewater samples*."

  medRxiv 2021.01.08.21249379; [doi:10.1101/2021.01.08.21249379](https://doi.org/10.1101/2021.01.08.21249379)

## Contacts

If you experience problems running the software:

- We encourage to use the [issue tracker on GitHub](https://github.com/cbg-ethz/cojac/issues)
- For further enquiries, you can also contact the [V-pipe Dev Team](https://cbg-ethz.github.io/V-pipe/contact/)
- You can contact the publication’s corresponding author
